File: LeetCode-Java/Longest Substring Without Repeating Characters.java
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s == null || s.length() == 0)
            return 0;
        HashMap<Character,Integer> record = new HashMap<Character,Integer>();
        int currentMax = 1;
        int result = 1;
        record.put(s.charAt(0), 0);
        for(int i = 1;i < s.length();i ++)
        {
            char val = s.charAt(i);
            int lastIndex = -1;
            if(record.containsKey(val))
                lastIndex = record.get(val);
            currentMax = Math.min(currentMax + 1, i - lastIndex);
            result = Math.max(currentMax, result);
            record.put(val, i);
        }
        return result;
    }
}

File: Leetcode-cpp/longest_substring_without_repeating_characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> record;
        int result = 0;
        int current_start = 0;
        for (int i = 0; i < s.length(); ++i) {
            if (record.find(s[i]) != record.end() && record[s[i]] >= current_start) {
                result = max(result, i - current_start);
                current_start = record[s[i]] + 1;
            }
            record[s[i]] = i;
        }
        result = max(result, (int)s.length() - current_start);
        return result;
    }
};